Blocks/Persistence/file.h

```diff
@@ -302,37 +345,68 @@ class FilePersister {
   template <typename E>
   idxts_t DoPublish(E&& entry, const std::chrono::microseconds timestamp) {
     end_t iterator = file_persister_impl_->end.load();
-    if (timestamp < iterator.us) {
-      CURRENT_THROW(ss::InconsistentTimestampException(iterator.us, timestamp));
+    if (!(timestamp > iterator.head)) {
+      CURRENT_THROW(ss::InconsistentTimestampException(iterator.head + std::chrono::microseconds(1), timestamp));
     }
-    iterator.us = timestamp;
-    const auto current = idxts_t(iterator.index, iterator.us);
+    iterator.last_entry_us = iterator.head = timestamp;
+    const auto current = idxts_t(iterator.next_index, iterator.last_entry_us);
     {
       std::lock_guard<std::mutex> lock(file_persister_impl_->mutex);
-      CURRENT_ASSERT(file_persister_impl_->offset.size() == iterator.index);
-      CURRENT_ASSERT(file_persister_impl_->timestamp.size() == iterator.index);
+      CURRENT_ASSERT(file_persister_impl_->offset.size() == iterator.next_index);
+      CURRENT_ASSERT(file_persister_impl_->timestamp.size() == iterator.next_index);
       file_persister_impl_->offset.push_back(file_persister_impl_->appender.tellp());
       file_persister_impl_->timestamp.push_back(timestamp);
     }
     file_persister_impl_->appender << JSON(current) << '\t' << JSON(std::forward<E>(entry)) << std::endl;
-    ++iterator.index;
-    iterator.us += std::chrono::microseconds(1);
+    ++iterator.next_index;
+    file_persister_impl_->head_offset = 0;
     file_persister_impl_->end.store(iterator);
     return current;
   }
 
-  bool Empty() const noexcept { return !file_persister_impl_->end.load().index; }
-  uint64_t Size() const noexcept { return file_persister_impl_->end.load().index; }
+  void DoUpdateHead(const std::chrono::microseconds timestamp) {
```

Comment:
You mean to explain why the mutex isn't used here? And I don't understand your words about "the two method in two classes". I thought I added `DoUpdateHead` and `UpdateHead` methods completely symmetrical to the existing `DoPublish` and `Publish` ones. Perhaps, I should look through that code again :)